File: java/examples/Driver.java
/**
 *  libcaca       Java bindings for libcaca
 *  Copyright (c) 2009 Adrien Grand <jpountz@dinauz.org>
 *
 *  This library is free software. It comes without any warranty, to
 *  the extent permitted by applicable law. You can redistribute it
 *  and/or modify it under the terms of the Do What The Fuck You Want
 *  To Public License, Version 2, as published by Sam Hocevar. See
 *  http://sam.zoy.org/wtfpl/COPYING for more details.
 */

import org.zoy.caca.Canvas;
import org.zoy.caca.Color;
import org.zoy.caca.Display;
import org.zoy.caca.Event;
import org.zoy.caca.TimeoutException;

public class Driver {

  public static void main(String[] args) {
    Canvas cv = new Canvas(32, 16);
    Display dp = new Display(cv);
    cv.setColor(Color.Ansi.WHITE, Color.Ansi.BLACK);
    Display.Driver driver;
    int i, cur = 0;
    while(true) {
      cv.put(1, 0, "Available drivers");
      driver = dp.getDriver();
      Display.Driver[] list = Display.getDriverList();
      for (i = 0; i < list.length; i++) {
        if (driver.equals(list[i])) {
          cur = i;
        }
        cv.drawLine(0, 2*i+2, 9999, 2*i+2, ' ');
        cv.put(2, 2*i+2, (cur == i ? "* " : "  ") + list[i].getCode() + " " + list[i].getDescription());
      }
      cv.put(1, 2*i + 2, "Switching driver in 5 seconds");
      dp.refresh();
      try {
        dp.getEvent(Event.Type.KEY_PRESS, 5000000);
        break;
      } catch(TimeoutException e) {
        // Let's continue
      }
      cur++;
      if (list[cur].getCode().equals("raw")) cur++;
      if (cur >= list.length) cur = 0;
      dp.setDriver(list[cur]);
    }
  }

}

File: python/examples/driver.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
#
# driver        libcaca test drivers program
# Copyright (c) 2010 Alex Foulon <alxf@lavabit.com>
#
# This file is a Python port of "examples/driver.c"
# which is:
# Copyright (c) 2006-2010 Sam Hocevar <sam@hocevar.net>
#               All Rights Reserverd
#
# This library is free software. It comes without any warranty, to
# the extent permitted by applicable law. You can redistribute it
# and/or modify it under the terms of the Do What the Fuck You Want
# to Public License, Version 2, as published by Sam Hocevar. See
# http://www.wtfpl.net/ for more details.
#

import sys

import caca
from caca.canvas import Canvas, CanvasError
from caca.display import Display, DisplayError, Event

def main():
    """ Main function. """

    lst = caca.get_display_driver_list()
    cur = 0

    try:
        cv = Canvas(0, 0)
        dp = Display(cv)
    except (CanvasError, DisplayError) as err:
        sys.stderr.write("%s\n" % err)
        sys.exit(127)

    cv.set_color_ansi(caca.COLOR_WHITE, caca.COLOR_BLACK)

    while True:
        cv.put_str(1, 0, "Available drivers:")
        cur_driver = dp.get_driver()
        n = 0
        for driver, desc in lst:
            if driver == cur_driver:
                cv.put_str(2, n + 2, "%s %s (%s)" % ('*', driver, desc))
            else:
                cv.put_str(2, n + 2, "%s %s (%s)" % (' ', driver, desc))
            n += 1

        cv.put_str(2, n + 3, "Switching driver in 5 seconds")
        dp.refresh()

        if dp.get_event(caca.EVENT_KEY_PRESS, Event(), 5000000):
            break

        cur += 1
        if cur < len(lst) and lst[cur][0] == "raw":
            cur += 1
        if cur >= len(lst):
            cur = 0

        dp.set_driver(lst[cur][0])

if __name__ == "__main__":
    main()

